Implement volumetric fog support for both point lights and spotlights

Soulghost opened the pull request:
# Objective
- Fixes: https://github.com/bevyengine/bevy/issues/14451

## Solution
- Adding volumetric fog sampling code for both point lights and spotlights.

## Testing
- I have modified the example of volumetric_fog.rs by adding a volumetric point light and a volumetric spotlight.

https://github.com/user-attachments/assets/3eeb77a0-f22d-40a6-a48a-2dd75d55a877

alice-i-cecile commented on crates/bevy_pbr/src/cluster/mod.rs:
```diff
@@ -505,12 +505,23 @@ impl Default for GpuClusterableObjectsUniform {
 pub(crate) fn clusterable_object_order(
-    (entity_1, shadows_enabled_1, is_spot_light_1): (&Entity, &bool, &bool),
-    (entity_2, shadows_enabled_2, is_spot_light_2): (&Entity, &bool, &bool),
+    (entity_1, shadows_enabled_1, is_volumetric_light_1, is_spot_light_1): (
```
Nit: This collection of bools is starting to get confusing; maybe refactor this into a private helper struct?

alice-i-cecile commented on crates/bevy_pbr/src/render/mesh_view_types.wgsl:
```diff
@@ -19,6 +19,7 @@ struct ClusterableObject {
 
 const POINT_LIGHT_FLAGS_SHADOWS_ENABLED_BIT: u32   = 1u;
 const POINT_LIGHT_FLAGS_SPOT_LIGHT_Y_NEGATIVE: u32 = 2u;
+const POINT_LIGHT_FLAGS_VOLUMETRIC_BIT: u32        = 4u;
```
Genuine question: why is this 4 rather than 3?

tychedelia replied:
`1 << 2` is `0b100` aka "the third flag" but this bit pattern is the literal `4u`. wgsl doesn't support defining consts via bitwise ops which is why you will sometime see flags like `const FLAG = 268435456` (2^28) or some other large number for high numbered flags.

tychedelia replied:
1 << 2 is 0b100 aka "the third flag" but this bit pattern is the literal 4u. wgsl doesn't support defining consts via bitwise ops which is why you will sometime see flags like const FLAG = 268435456 (2^28) or some other large number for high numbered flags.

alice-i-cecile commented on examples/3d/volumetric_fog.rs:
```diff
@@ -60,6 +61,44 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
+                intensity: 1000.0,
+                ..default()
+            },
+            transform: Transform::from_xyz(-0.3493744, 1.900556, 1.0452124),
```
Nit: can we pick some slightly rounder numbers to make reading the example easier?

alice-i-cecile commented on examples/3d/volumetric_fog.rs:
```diff
@@ -60,6 +61,44 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
+        .insert(VolumetricLight);
+
+    // Add the spot light
+    let mut spotlight_transform = Transform::from_xyz(-1.7817883, 3.901562, -2.7141085);
```
When defining scenes via code for examples like this, I often find `Transform::look_at` to be much easier to setup intent.

alice-i-cecile reviewed (commented):
Nothing blocking, but a few questions and suggestions. I'll leave the full reviews to the rendering experts, but I'm very pleased to see this!

tychedelia commented on examples/3d/volumetric_fog.rs:
```diff
@@ -60,6 +61,44 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
             ..default()
         });
 
+    // Add the point light
```
It would be nice to add a UI option to cycle whether the lights are enabled, so that users can still see the raw effect.

Soulghost commented:
Hi @alice-i-cecile and @tychedelia , thank you for the guidance. I have refactored the code and added some UI options for the volumetric fog example. Could you please help review this again? Thank you.
<img width="1280" alt="image" src="https://github.com/user-attachments/assets/78ac5039-b3b4-4f20-9206-59c47adae34d">

IceSentry commented on crates/bevy_pbr/src/render/light.rs:
```diff
@@ -616,6 +634,14 @@ pub fn prepare_lights(
+    let point_light_volumetric_enabled_count = point_lights
+        .iter()
+        .filter(|(_, light, _)| {
+            (light.volumetric || light.shadows_enabled) && light.spot_light_angles.is_none()
```
Why do you need the shadows_enabled here?

Soulghost replied:
To ensure that the volumetric spot light is activated correctly after the non-volumetric, shadow enabled point light. A better approach would be to subtract `point_light_count ` instead. Let me optimize the code, thank you.

IceSentry commented on examples/3d/volumetric_fog.rs:
```diff
@@ -60,6 +81,37 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
 
+    // Add the point light
+    commands
+        .spawn(PointLightBundle {
```
Would be nice to be able to move the point light around. Just moving statically between back and forth between the walls would probably be enough. Not really a blocker for the PR, just something that would be nice.

Soulghost replied:
The point light now moves back and forth between the left and right walls.

https://github.com/user-attachments/assets/3eeb77a0-f22d-40a6-a48a-2dd75d55a877

IceSentry reviewed (approved):
Everything looks good to me. I didn't dive too deep into the math, but it looks like it works and nothing is obviously wrong.

tychedelia reviewed (approved):
Verified all the fog examples work.

alice-i-cecile commented:
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1702 if you'd like to help out.
